envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/JwtFilterFactory.kt

```diff
@@ -0,0 +1,103 @@
+import pl.allegro.tech.servicemesh.envoycontrol.groups.IncomingEndpoint
+import pl.allegro.tech.servicemesh.envoycontrol.groups.OAuth
+import pl.allegro.tech.servicemesh.envoycontrol.snapshot.JwtFilterProperties
+import pl.allegro.tech.servicemesh.envoycontrol.snapshot.OAuthProvider
+
+class JwtFilterFactory(
+    private val properties: JwtFilterProperties
+) {
+
+    fun createJwtFilter(group: Group): HttpFilter? {
+        return if (shouldCreateFilter(group)) {
+            HttpFilter.newBuilder()
+                .setName("envoy.filters.http.jwt_authn")
+                .setTypedConfig(
+                    Any.pack(
+                        JwtAuthentication.newBuilder().putAllProviders(
+                            getOAuthProviders()
+                        ).addAllRules(
+                            createRules(group.proxySettings.incoming.endpoints)
+                        )
+                            .build()
+                    )
+                )
+                .build()
+        } else {
+            null
+        }
+    }
+
+    private fun shouldCreateFilter(group: Group): Boolean {
+        return group.proxySettings.incoming.endpoints.any { it.oauth != null } && properties.providers.isNotEmpty()
+    }
+
+    private fun getOAuthProviders(): Map<String, JwtProvider> =
+        properties.providers.associate {
+            it.name to createProvider(it)
+        }
+
+    private fun createProvider(provider: OAuthProvider) = JwtProvider.newBuilder()
+        .setRemoteJwks(
+            RemoteJwks.newBuilder().setHttpUri(
+                HttpUri.newBuilder()
+                    .setUri(provider.jwksUri.toString())
+                    .setCluster(provider.clusterName)
+                    .setTimeout(
+                        Durations.fromMillis(provider.connectionTimeout.toMillis())
+                    ).build()
+            )
+                .setCacheDuration(Durations.fromMillis(provider.cacheDuration.toMillis()))
+        )
+        .setIssuer(provider.name)
+        .setForward(true)
+        .setForwardPayloadHeader(properties.forwardPayloadHeader)
+        .setPayloadInMetadata(properties.payloadInMetadata)
+        .build()
+
+    // without config for now
+    private fun createRules(endpoints: List<IncomingEndpoint>): List<RequirementRule> {
+        return endpoints.mapNotNull(this::createRuleForEndpoint)
+    }
+
+    private fun createRuleForEndpoint(endpoint: IncomingEndpoint): RequirementRule? {
+        return if (endpoint.oauth != null) {
+            RequirementRule.newBuilder().setMatch(
+                RouteMatch.newBuilder().setPrefix(endpoint.path)
+            ).setRequires(
+                createJwtRequirement(endpoint.oauth.provider, endpoint.oauth.policy)
+            ).build()
+        } else {
+            null
+        }
+    }
+
+    private fun createJwtRequirement(provider: String, policy: OAuth.Policy): JwtRequirement {
+        return when (policy) {
```

Comment:
Not sure if that helps a lot but we could make a map Map<Key(provider,policy), JwtRequirement> and initialize all at the beginning. Because, provider in node metadata need to be equals to one in properties we could make all of them and later just retrieve them from map.

Reply:
done